### ISSUE TITLE
Type synonyms for Haskell defined types

### ISSUE THREAD

**Issue by bosu** [Gabriel439/Haskell-Dhall-Library#92]
I'd like to let the users write lambdas like:

`λ(x : Big) -> ...`

where `Big` is the type synonym for some big (many fields) record or union type. This record data type will be generated from the Haskell code. It would be very inconvenient to make the users write it manually.

Unfortunately I do not see the way to achieve this.

Some of the changes I've thought about:

1. Adding new `TypeDef` constructor to `Expr`
2. Passing `Context.lookup` to `typeWith` and overloading it to return the record instead of `Big`
3. Having `typeWith` work on some kind of `Context (Either Synonym Expr)` data type.
4. New `FromContext Text` `PathType` constructor allowing replacing the types during `loadWith`.
5. Somethings obvious that I am missing? :)

**Comment by Gabriel439** [comment]
I assume that you've already read https://github.com/Gabriel439/Haskell-Dhall-Library/issues/10, but I'll still mention it in case somebody else comes across this issue and doesn't have that context

There's one other option, which is that you host your type at some URL (i.e. `https://ipfs.io/ipfs/.../Big`) and then ask your users to "install" the large type like this:

```bash
$ echo 'https://ipfs.io/ipfs/.../Big' > ./Big
```

This lets users write:

```bash
λ(x : ./Big) → ...
```

The reason Dhall does not have a `type` synonym construct like Haskell is that this would require interleaving evaluation and type-checking.  For example, consider this hypothetical code using such a construct:

```bash
    let  g = λ(t : Type → Type) → t
in  type L = g List
in  let  f = λ(a : Type) → L (L a)
in  type Example = f Integer
in  λ(x : Example) → ...
```

That would require multiple rounds of normalization and type-checking

The only reason Haskell gets away with having type synonyms is because Haskell types are not first class values and Haskell type synonyms are (even more so) not first-class values, since there are many places in Haskell where a type is valid and a type synonym is not

**Comment by bosu** [comment]
> I assume that you've already read #10, but I'll still mention it in case somebody else comes across this issue and doesn't have that context

Actually I have not. Thanks for mentioning it. I do agree with your reasoning.

I guess what bothers me is the need to have these generated types as the files on the disk. It is very inconvenient during testing :(.

Would you be open to refactoring of Dhall.Import allowing pure `load` function? I am thinking along the
lines of having something like:

`
load' :: MonadCatch m => (Path -> m (Expr Src Path)) -> Expr Src Path -> m (Expr Src X)
`

Then using a custom `Path` resolving callback with `ExceptT` based monad will solve my issues.
If yes, I am willing to send a pull request.

**Comment by bosu** [comment]
Please see https://github.com/Gabriel439/Haskell-Dhall-Library/pull/93.

**Comment by Gabriel439** [comment]
Sorry for the delay as I was on vacation.  I'm still thinking if there is a better solution to this problem or not.  Give me a few days

**Comment by Gabriel439** [comment]
Yeah, after more thought I think this is the right approach.  I'll add more comments to the corresponding pull request

**Comment by Gabriel439** [comment]
Is this fixed by #93?

**Comment by bosu** [comment]
Yes, works for me.
